File: backend/app/utils/privacy.py
"""
Server-side coordinate fuzzing for public (unauthenticated) API responses.

Uses the same djb2-based algorithm as frontend/src/utils/privacy.ts so
markers appear in the same fuzzed position regardless of which side renders.
"""


def _hash_code(s: str) -> int:
    """Deterministic 32-bit djb2 hash matching the TypeScript implementation."""
    h = 5381
    for c in s:
        # Replicate JS signed 32-bit integer overflow behaviour
        h = ((h << 5) + h + ord(c)) & 0xFFFFFFFF
        # Convert to signed 32-bit (JS `| 0`)
        if h >= 0x80000000:
            h -= 0x100000000
    return h


def fuzz_coords(lat: float, lon: float, node_id: str) -> tuple[float, float]:
    """
    Apply a deterministic ±0.005° offset (~500 m) derived from the node ID.
    The offset is stable for a given ID so markers don't jump around.
    """
    h = _hash_code(node_id)
    # Replicate JS unsigned right-shift (`>>> 16`) — mask to 32-bit first
    h_unsigned = h & 0xFFFFFFFF
    lat_offset = ((h_unsigned & 0xFFFF) / 0xFFFF) * 0.01 - 0.005
    lon_offset = (((h_unsigned >> 16) & 0xFFFF) / 0xFFFF) * 0.01 - 0.005
    return lat + lat_offset, lon + lon_offset

File: frontend/src/utils/privacy.ts
/** Deterministic 32-bit hash (djb2). */
function hashCode(str: string): number {
  let hash = 5381
  for (let i = 0; i < str.length; i++) {
    hash = ((hash << 5) + hash + str.charCodeAt(i)) | 0
  }
  return hash
}

/**
 * Apply a deterministic random offset to coordinates (~500 m).
 * The offset is stable for a given id so markers don't jump around.
 */
export function fuzzCoords(lat: number, lon: number, id: string): { lat: number; lon: number } {
  const h = hashCode(id)
  // Two independent seeds from the same hash
  const latOffset = ((h & 0xffff) / 0xffff) * 0.01 - 0.005        // [-0.005, +0.005]
  const lonOffset = (((h >>> 16) & 0xffff) / 0xffff) * 0.01 - 0.005
  return { lat: lat + latOffset, lon: lon + lonOffset }
}

/** Format a coordinate value — masked when privacy mode is on. */
export function formatCoord(value: number, privacyOn: boolean): string {
  return privacyOn ? '*****' : value.toFixed(5)
}
